Use generics for workqueue metrics

antoninbas opened the pull request:
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

The workqueue implementation was recently updated to be strongly typed, using Go generics. However the metrics implementation was not updated, and continued using interface{}. This translated to unnecessary memory allocations when invoking the queueMetrics interface methods to track queue operation. We can avoid these extra heap allocations by using generics for the metrics implementation as well.

#### Which issue(s) this PR fixes:

Fixes #

#### Special notes for your reviewer:

See comment below for benchmark results before and after the change.

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

antoninbas committed:
Use generics for workqueue metrics

The workqueue implementation was recently updated to be strongly typed,
using Go generics. However the metrics implementation was not updated,
and continued using interface{}. This translated to unnecessary memory
allocations when invoking the queueMetrics interface methods to track
queue operation. We can avoid these extra heap allocations by using
generics for the metrics implementation as well.

Signed-off-by: Antonin Bas <antonin.bas@broadcom.com>

antoninbas commented:
I used the following benchmarks to identify the issue and validate the patch:
```go
package main

import (
	"fmt"
	"testing"

	"k8s.io/client-go/util/workqueue"
)

type sX struct {
	X int64
	Y int64
	Z int64
}

var keysStr []string
var keysPtr []*sX
var keysStruct []sX

func init() {
	for j := 0; j < 100; j++ {
		keysStr = append(keysStr, fmt.Sprintf("key-%d", j))
		keysPtr = append(keysPtr, &sX{})
		keysStruct = append(keysStruct, sX{
			X: int64(j),
		})
	}
}

func BenchmarkPointers(b *testing.B) {
	for i := 0; i < b.N; i++ {
		b.StopTimer()
		q := workqueue.NewTypedDelayingQueueWithConfig(workqueue.TypedDelayingQueueConfig[*sX]{})
		b.StartTimer()
		for j := 0; j < 100; j++ {
			q.Add(keysPtr[j])
			key, _ := q.Get()
			q.Done(key)
		}
	}
}

func BenchmarkStrings(b *testing.B) {
	for i := 0; i < b.N; i++ {
		b.StopTimer()
		q := workqueue.NewTypedDelayingQueueWithConfig(workqueue.TypedDelayingQueueConfig[string]{})
		b.StartTimer()
		for j := 0; j < 100; j++ {
			q.Add(keysStr[j])
			key, _ := q.Get()
			q.Done(key)
		}
	}
}

func BenchmarkStructs(b *testing.B) {
	for i := 0; i < b.N; i++ {
		b.StopTimer()
		q := workqueue.NewTypedDelayingQueueWithConfig(workqueue.TypedDelayingQueueConfig[sX]{})
		b.StartTimer()
		for j := 0; j < 100; j++ {
			q.Add(keysStruct[j])
			key, _ := q.Get()
			q.Done(key)
		}
	}
}
```

Results before applying the patch:
```
cpu: Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
BenchmarkPointers-16    	   63361	     19723 ns/op	    1268 B/op	     105 allocs/op
BenchmarkStrings-16     	   34248	     33197 ns/op	    7177 B/op	     407 allocs/op
BenchmarkStructs-16     	   30970	     38855 ns/op	   10500 B/op	     407 allocs/op
```

Results after applying the patch:
```
cpu: Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
BenchmarkPointers-16    	   64141	     18835 ns/op	    1239 B/op	     105 allocs/op
BenchmarkStrings-16     	   60921	     21467 ns/op	    2212 B/op	     105 allocs/op
BenchmarkStructs-16     	   47352	     25214 ns/op	    3167 B/op	     105 allocs/op
```

```
$ go version
go version go1.23.0 darwin/amd64
```

Notice that there is no extra heap allocation when casting a pointer to `interface{}`, and the results for `BenchmarkPointers` are the same before and after (this is expected: we already have a data pointer and don't need to escape to the heap). Workqueues in K8s and in K8s controllers typically use a string or a small struct.

tnqn commented:
/ok-to-test
/lgtm

It seems this optimization could also be applied to `waitFor`, which is the last remaining struct that doesn't yet use generics in this package, aside from this one. But it could be done in a separate patch if this optimization makes sense to the workqueue maintainers.
https://github.com/kubernetes/kubernetes/blob/9770283c137e73169a8e6f94abdf3e15f1f8cc4d/staging/src/k8s.io/client-go/util/workqueue/delaying_queue.go#L175-L180

tnqn commented:
Somehow the link checks think the updated functions are unused.

antoninbas commented:
> Somehow the lint checks think the updated functions are unused.

Yes, I believe this is because of a bug in the `staticcheck` linter, that has been fixed, but k8s currently uses an older version of golangci-lint. I may need to open an other PR first to bump golangci-lint.

cici37 commented:
/assign @aojea 
If you would have time for this one or help with pointing the right person. Thank you :)
/triage accepted

antoninbas commented:
@aojea will you have time to take a quick look at this PR?

aojea commented on staging/src/k8s.io/client-go/util/workqueue/metrics.go:
```diff
@@ -27,10 +27,10 @@ import (
-	get(item T)
-	done(item T)
-	updateUnfinishedWork()
+	Add(item T)
```
is the false positive blocking to merge?
if affirmative, it seems it is waiting for golang to fix it, can not solve this problem without having to modify the methods?

antoninbas replied:
Yes it blocks the merge.
There are potentially 3 options:
1. make the interface methods public while keeping the interface private, which is what I chose here, because it didn't seem like a big drawback
2. add nolint pragmas, probably in several places. However, https://github.com/dominikh/go-tools/issues/1294#issuecomment-2066322311 seems to indicate that this won't work.
3. wait, but to me the benefits of the PR outweigh the inconvenience of making these methods public. There is no timeline for https://github.com/golang/go/issues/63982 to be resolved

aojea replied:
I'm surprised this can not be solved with an exclude rule https://github.com/kubernetes/kubernetes/blob/d99d3f7eb734dc9284a90374fcc6d8cc32d1946d/hack/golangci.yaml#L37 , if the linter unused does not run in this file there should be no problem

aojea commented on staging/src/k8s.io/client-go/util/workqueue/metrics.go:
```diff
@@ -210,37 +210,24 @@ func (_ noopMetricsProvider) NewRetriesMetric(name string) CounterMetric {
 
-	onlyOnce sync.Once
-}
+var setGlobalMetricsProviderOnce sync.Once
```
this part is contentious https://github.com/kubernetes/kubernetes/issues/127739

cc: @alvaroaleman @dgrisonnet

antoninbas replied:
@alvaroaleman is correct. The change in "style" (removal of the `queueMetricsFactory` factory) is because a struct cannot declare type parameters (generics) that are specific to the method. So we cannot have a non-generic factory struct (necessary to declare a global variable) with methods that use generics.

antoninbas commented on staging/src/k8s.io/client-go/util/workqueue/metrics.go:
```diff
@@ -26,10 +26,10 @@ import (
 // of metrics.
 
-type queueMetrics interface {
-	add(item t)
```
Yes, a plain `any` would have been clearer in the original code. If the suggestion in https://github.com/kubernetes/kubernetes/pull/127635#issuecomment-2375858216 is implemented as a follow-up, I think `t` can be removed completely.

aojea commented on staging/src/k8s.io/client-go/util/workqueue/queue.go:
```diff
@@ -209,7 +205,7 @@ type Typed[t comparable] struct {
 	drain        bool
 
-	metrics queueMetrics
+	metrics queueMetrics[t]
```
just for reference, this lowercase `t` is the generic  type`comparable` , not the defined `type` `interface{}`

aojea commented:
+1 for the first commit but not for the second commit.

I don't like the idea to change the code for a linter/golang error https://github.com/kubernetes/kubernetes/pull/127635#discussion_r1786781287 and I think is better to exclude the `unused` linter on that file (don't know if it works though but we should check that option)

antoninbas commented:
> +1 for the first commit but not for the second commit.
> 
> I don't like the idea to change the code for a linter/golang error [#127635 (comment)](https://github.com/kubernetes/kubernetes/pull/127635#discussion_r1786781287) and I think is better to exclude the `unused` linter on that file (don't know if it works though but we should check that option)

Thanks for the review.
No problem for the requested change on my side. I will try it and update the PR if it works.

aojea commented:
> No problem for the requested change on my side. I will try it and update the PR if it works.

thanks,  please add a comment with the context and the references, so future us or new contributors remove the exception once the problem is fixed

antoninbas committed:
Ignore false positives from unused linter

This is a workaround for a false positive in the staticcheck linter,
which has not been addressed yet.
See https://github.com/dominikh/go-tools/issues/1294.
Once this is fixed in staticcheck, we can just remove the exclude rule
from the golangci config.

Signed-off-by: Antonin Bas <antonin.bas@broadcom.com>

antoninbas commented:
@aojea I have updated the golangci config as we discussed, and updated the second commit accordingly.

BTW, I have also tried adding a `//nolint:unused` directive on individual methods, and it also seems to be working, assuming that the directive is added on all the methods for `noMetrics[T]` and `defaultQueueMetrics[T]` (9 in total). I do believe that the golangci config solution is better, as it gives us one central place where we can add a comment with the necessary context, and extra files can be added in the future if other parts of the K8s codebase run into the same issue. Let me know what you think.

antoninbas commented:
/retest

aojea commented:
/lgtm

/assign @sttts @liggitt 

for approval

sttts commented:
/approve
